### DISCUSSION TITLE
Block AI crawlers

### DISCUSSION THREAD

**Discussion by HeuvelTop** [fail2ban/fail2ban#3772] [Ideas]
Last week I have been trying to defend my web sites of article stealing AI-crawlers. Because, obviously, they don't like to follow "robots.txt" restrictions.

So I've created the following file in filter.d: apache-crawler.conf:
_{some updates from below comments are implemented}_

```
[Definition]
_bots = [^"]+(?:b.?o.?t|crawler|sp[iy]der|008|accoona|arachmo|archive\.org|baidu|cerberian|charlotte|collage|context_ad|cosmos|covario|crawl|copier|curl|datadome|dataparksearch|downloader|earthcom|EmailSiphon|emeraldshield\.com|explorer|facebo|findlinks|holmes|htdig|ia_archiver|ichiro|webis|larbin|libwww-perl|linkwalker|lwp-trivial|mabontland|masscan\-ng|google|mogi|morning\spaper|mvaclient|netresearchserver|newsgator|nikto|nmap|nutchcvs|nymesis|oegp|orbiter|peew|pompos|postpost|pycurl|python-requests|qseero|radian6|sbider|scoutjet|scrapy|scrubby|search|semanticdiscovery|shopwiki|silk|slurp|snappy|sqlmap|squash|sqworm|stackrambler|teoma|tineye|truwogps|updated|vagabondo|vortex|voyager|vyu2|webcollage|websquash|wf84|wget|wofindeich|womlpefactory|yacy|yahoo|yandex|yasakli|yeti|yodao|yooglifetchagent|youdao|zao|zealbot|zoominfo|zyborg)[^"]*
_req = GET\s[^"]*
_referrer = -
failregex = ^<ADDR>[^"]+"%(_req)s" \d+ \d+ "%(_referrer)s" "[^"]*%(_bots)s\b
```
In practice the _req can be written to implement the rules from robots.txt, basically making it only punish the crawlers that ignore the rules.

And then add the following to jail.local:
```
[apache-crawler]
enabled = true
filter = apache-crawler
logpath = /var/log/apache2/*access.log   # Adjust this path to your Apache access log
maxretry = 1                           # no patience here
bantime = 600                        # If recidive is not active a longer Ban time may be appropriate
findtime = 600                        # Find time in seconds (e.g., 600 for 10 minutes)
action = iptables[name=apache-crawler, port=http, protocol=tcp]
```

Now this already seems to add quite some IP's to the block list. I would really like comments and if possible additions, corrections.

Addition (on 27-6-2024): here is a site that lists user-agent-id's that may help to extend the above regex : https://useragentstring.com/pages/useragentstring.php. It needs some testing to find the best strategy. I have changed the _bots list heavily and also implemented an ignoreregex to exclude some benign ones.

Warmly,
Frank

**Comment by sebres** [comment]
1. Your regex is vulnerable - not anchored, contains catch-alls etc. Thus it can theoretically match legitimate attempts as well as being an issue of potentially affected by inject foreign data, for instance, imagine your competitor will put 3 URLs on his web-side that would contain `inject=HTTP...bot` in URI (as parameter or part of URI-path) - this will cause that people firstly visited his site before yours will be unable to visit your web-site, because become banned (this is just an example, the "attack"-vectors may be more complex).
Not to mention that this RE is very slow and expensive.
Anyway `failregex` must be as precise as possible (e. g. list of blacklisted bots must definitely match user-agent and not some part of URI), anchored and shall not contain catch-alls.
2. Ban for 1 day is too hard, better would be smaller initial `bantime` and `bantime.increment = true`, so it'd ban recidive evildoers for longer time (and faster) if they'd repeat their attempts after ban.
3. Banning bots by user-agent is also not good idea: it is pretty simple to change user-agent, so you'd never consider them (or even see some browser names or good bots instead). There are many approaches to restrict bots, for example - limiting requests and/or connections and then ban bots violating the limits.
4. Determining whether a bot is legal bot or not is also not trivial task, e.g. for googlebot - reverse DNS lookup (see https://developers.google.com/search/docs/crawling-indexing/verifying-googlebot) etc.
5. Monitoring of access-log is basically bad idea - see [wiki :: Best-practice](https://github.com/fail2ban/fail2ban/wiki/Best-practice), especially part "Reduce parasitic log-traffic".
For instance in your case, you'd have some conditional logging (for bots touching locations outside of `robots.txt` only) and monitor this log only.

**Reply by HeuvelTop** [reply]
Thanks. Also for your time in making these comments. I'll first take your fifth point to improve this method.

I also have recidive on, so shortening the bantime will be done.

One of my web sites publishes a lot of articles that we don't want to be just taken by AI-crawlers, especially when they just ignore the robots.txt file. And separating the log files is probably the easiest to implement.

Then I'll try to improve the regex. But I'll need to collect logging samples to test this.

I'll come back when I managed to improve it.

**Reply by sebres** [reply]
> Then I'll try to improve the regex. But I'll need to collect logging samples to test this.

Well, simple "improvement" is to anchor it, to remove catch-alls and to consider the bot name only in user-agent field...
Something like this:
```ini
_bots = (?:Botname1|Botname2|...|BotnameN)
_req = [^"]*
_referrer = [^"]*
failregex = ^<ADDR>[^"]+"%(_req)s" \d+ \d+ "%(_referrer)s" "%(_bots)s\b
```
(Note: `\b` is used as word-boundary for names of bots, so with `Bot\b` it'd match `Bot/1.02` or `Bot-XYZ`, but not `Bottle`, this `_bots` is a blacklist of prefixes, considering end of word).

However this whole RE surely depends on the log-format used for logging.
For example it would match this log-line:
```
192.0.2.123 - - ...no matter what is here (excepting a quote-char)... "HEAD /some-url/ HTTP/1.0" 403 459 "https://example.tld/referrer-url/" "Botname1/1.02"'
```
But no matter how you'd adjust it, try to avoid catch-alls, e. g. see both vars `_req` and `_referrer` which use `[^"]*`, what means everything excepting `"`-character (quote must be url-encoded as `%22`), so it'd match to end of requested URI with protocol or to end of referrer up top closing quote-char and match further with rest of RE, e. g. one of the bot prefixes.
This will make the RE more precise, safer and many times faster.

**Reply by HeuvelTop** [reply]
Thanks for your input. I have been looking at the log files. I am using combined LogFormat:
`
  "%h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-agent}i\""
`
The User-Agent part does not match with the regex you suggested. I have some example user-agent strings from crawlers here:
```
"Mozilla/5.0 (Linux; Android 6.0.1; Nexus 5X Build/MMB29P) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.6422.175 Mobile Safari/537.36 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)"
"Mozilla/5.0 AppleWebKit/537.36 (KHTML, like Gecko; compatible; bingbot/2.0; +http://www.bing.com/bingbot.htm) Chrome/116.0.1938.76 Safari/537.36"
"Mozilla/5.0 (compatible; SemrushBot/7~bl; +http://www.semrush.com/bot.html)"
```

Also it still is true that all the log lines from crawlers have an empty referrer. So I suggest the following:
```
_bots = (?:Botname1|Botname2|...|BotnameN)
_req = GET\s[^"]*
_referrer = -
failregex = ^<ADDR>[^"]+"%(_req)s" \d+ \d+ "%(_referrer)s" "[^"]*%(_bots)s\b
```
Now I can move on to designing conditional logging so only the ones who violate my robots.txt will be locked out.

**Reply by HeuvelTop** [reply]
After some intensive testing, it turns out it is more easy to write a regex for the request (_req) to match the disallow statements in robots.txt. The funfact is that crawlers who obey robots.txt don''t get banned with that addition.
With some careful inspection of the log lines I found that "facebookexternalhit" and "semrushbot" specifically mess with URL's and try all kinds of made up url's.
This is an example of what you can do with the request part of the setup:
`
_req = GET[^"]+(?:/\?|/pathname1/|/pathname2/)[^"]*
`
Where the /? is part of the facebook mess and the pathname1 and 2 are examples of what you have disallowed in robots.txt.

**Comment by HeuvelTop** [comment]
From running the filter on log files of last weeks I learned that:
- SemrushBot, and bingbot both mess with the URL's creating very long repetative pathname, possibly in an attempt to evade filters.
- facebookexternalhit tries to make up URL's starting with /?search or other variants starting with /?.

IMHO these should go into the badbots filter.

**Reply by mantaalex** [reply]
thanks for sharing your filter i had some trouble blocking Claudebot but i now managed to add it to the list and block it thanks !
